### PR TITLE
Ironic agent with swift

### PR DESCRIPTION
**Pull request by skazi0**
Ironic agent_* drivers require proper Swift configuration for downloading
images. Additional validations were added to Ironic barclamp to make sure
all requirements are in place when using agent_* drivers.

Ironic uses TempURL feature of Swift for providing images to deployment
agent. Random TempURL key is set for glance user if it was not defined
before.

Ironic barclamp was modified to expose enabled_drivers as an attribute.

### CODE REVIEW

**Comment by skazi0** [comment]
[(3) S60P6 Integrate agent_* deployment drivers for Ironic](https://trello.com/c/DPgQWTr0/526-3-s60p6-integrate-agent-deployment-drivers-for-ironic)

File: chef/cookbooks/ironic/recipes/server.rb [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+openstack_command << " --os-password #{keystone_settings["admin_password"]}"
+openstack_command << " --os-tenant-name #{keystone_settings["admin_tenant"]}"
+openstack_command << " --os-auth-url #{auth_url} #{insecure}"
+
+get_ironic_net_id = "#{openstack_command} network show ironic -f value -c id"
+ironic_net_id = Mixlib::ShellOut.new(get_ironic_net_id).run_command.stdout.chomp
+
+swift = search(:node, "roles:swift-proxy").first || {}
+# configure swift only if some agent_* drivers are enabled
+if swift && !node[:ironic][:enabled_drivers].select { |d| d.start_with?("agent_") }.empty?
```
Comment (jsuchome) [inline chef/cookbooks/ironic/recipes/server.rb]: What about
`if swift && node[:ironic][:enabled_drivers].select { |d| d.start_with?("agent_") }.any?`

File: chef/cookbooks/ironic/recipes/server.rb [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+  glance_keystone_settings = KeystoneHelper.keystone_settings(glance, "glance")
+
+  swift_command = "swift --os-username #{glance_keystone_settings["service_user"]}"
+  swift_command << " --os-password #{glance_keystone_settings["service_password"]}"
+  swift_command << " --os-tenant-name #{glance_keystone_settings["service_tenant"]}"
+  swift_command << " --os-identity-api-version 3 --os-auth-url #{auth_url}"
+  swift_command << (glance_keystone_settings["insecure"] ? " --insecure" : "")
```
Comment (jsuchome) [inline chef/cookbooks/ironic/recipes/server.rb]: Shouldn't you (additionally) check for swift insecure settings?

File: chef/cookbooks/ironic/recipes/server.rb [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    tempurl_key = secure_password
+    execute "set-glance-tempurl-key" do
+      command "#{swift_command} post -m 'Temp-Url-Key:#{tempurl_key}'"
+      user glance[:glance][:user]
+      group glance[:glance][:group]
+    end
+  end
+
+  # use IP as this will be used by agent which can have no DNS configured
+  swift_address = Chef::Recipe::Barclamp::Inventory.get_network_by_type(swift, "public").address
```
Comment (jsuchome) [inline chef/cookbooks/ironic/recipes/server.rb]: `Barclamp::Inventory.get_network_by_type`

File: chef/cookbooks/ironic/recipes/server.rb [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+  # swift_address = CrowbarHelper.get_host_for_admin_url(swift, swift[:swift][:ha][:enabled]),
```
Comment (jsuchome) [inline chef/cookbooks/ironic/recipes/server.rb]: ??

File: crowbar_framework/app/models/ironic_service.rb [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    validate_one_for_role proposal, "ironic-server"
+
+    net_svc = NetworkService.new @logger
+    network_proposal = Proposal.find_by(barclamp: net_svc.bc_name, name: "default")
+    if network_proposal["attributes"]["network"]["networks"]["ironic"].nil?
+      validation_error I18n.t("barclamp.#{@bc_name}.validation.ironic_network")
+    end
+
+    # additional soft dependencies for agent_* drivers
+    unless proposal["attributes"][@bc_name]["enabled_drivers"].select {|d| d.start_with?("agent_")}.empty?
```
Comment (jsuchome) [inline crowbar_framework/app/models/ironic_service.rb]: Don't we need validation for other kinds of enabled drivers?

File: chef/cookbooks/ironic/recipes/server.rb (hunk 1 above)
Comment (skazi0) [inline chef/cookbooks/ironic/recipes/server.rb]: Oh, with `any?` we can do even better.

File: crowbar_framework/app/models/ironic_service.rb (hunk 5 above)
Comment (skazi0) [inline crowbar_framework/app/models/ironic_service.rb]: No that I know of.

File: chef/cookbooks/ironic/recipes/server.rb (hunk 2 above)
Comment (skazi0) [inline chef/cookbooks/ironic/recipes/server.rb]: I switched to using swift's insecure setting. I think this should be enough.

Comment (jsuchome) [inline chef/cookbooks/ironic/recipes/server.rb]: I actually thought that we might need to check for both. ... but swift is probably enough

**Comment by skazi0** [comment]
rebased

**Comment by skazi0** [comment]
rebased

**Comment by skazi0** [comment]
rebased

**Pull request merged**